sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java

```diff
@@ -25,40 +24,50 @@
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
 
 /**
  * Processes AMQP receive links into a stream of AMQP messages.
  */
 public class AmqpReceiveLinkProcessor extends FluxProcessor<AmqpReceiveLink, Message> implements Subscription {
-    // We don't want to dump too many credits on the link at once. It's easy enough to ask for more.
-    private static final int MINIMUM_REQUEST = 0;
-    private static final int MAXIMUM_REQUEST = 100;
-
     private final ClientLogger logger = new ClientLogger(AmqpReceiveLinkProcessor.class);
     private final Object lock = new Object();
     private final AtomicBoolean isTerminated = new AtomicBoolean();
-    private final AtomicBoolean hasDownstream = new AtomicBoolean();
     private final AtomicInteger retryAttempts = new AtomicInteger();
-    private final AtomicInteger linkCreditRequest = new AtomicInteger(1);
     private final Deque<Message> messageQueue = new ConcurrentLinkedDeque<>();
+    private final AtomicBoolean hasFirstLink = new AtomicBoolean();
+    private final AtomicBoolean linkCreditsAdded = new AtomicBoolean();
 
     private final int prefetch;
     private final AmqpRetryPolicy retryPolicy;
-    private Disposable parentConnection;
-
-    private volatile Subscription upstream;
-    private volatile CoreSubscriber<? super Message> downstream;
-    private volatile boolean isCancelled;
+    private final Disposable parentConnection;
 
     private volatile Throwable lastError;
+    private volatile boolean isCancelled;
     private volatile AmqpReceiveLink currentLink;
     private volatile Disposable currentLinkSubscriptions;
     private volatile Disposable retrySubscription;
 
-    volatile int wip;
-    static final AtomicIntegerFieldUpdater<AmqpReceiveLinkProcessor> WIP =
+    private volatile CoreSubscriber<? super Message> downstream;
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<AmqpReceiveLinkProcessor, CoreSubscriber> DOWNSTREAM =
```

Comment:
Can we use `AtomicReference` here instead of `AtomicReferenceFieldUpdater`? I don't think we'll have too many instances of AmqpReceiveLinkProcessor for memory being an issue. The reflection-based update can cause issues if the name of the volatile var is changed and the field updater is not updated.

Reply:
Makes sense. I was thinking of those EHs that have 1k partitions. I can revert it.